### PR TITLE
Improving asyncOp_sendmany redundancies (first round)

### PR DESCRIPTION
**Pull request by furszy**
Work over async sendmany operation flow. Specifically on the utxo finding, utxo selection, total amount calculation and the tx inputs appending process.  Removed several unneeded for loops over the entire utxo set, re structuring and cleaning the flow.

### CODE REVIEW

**Comment by rex4539** [comment]
Rebased to master and all tests pass.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 1]
```diff
@@ -913,9 +857,64 @@ bool AsyncRPCOperation_sendmany::find_utxos(bool fAcceptCoinbase=false) {
...
 
+    // Load transparent inputs
+    load_inputs(txValues);
+
     return t_inputs_.size() > 0;
 }
 
+bool AsyncRPCOperation_sendmany::load_inputs(TxValues& txValues) {
+    // If from address is a taddr, select UTXOs to spend
+    CAmount selectedUTXOAmount = 0;
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:863]: Do we need `txValues.selectedUTXOCoinbase = false;` here?

**Review by daira — APPROVED** [review]
utACK with minor comments.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 2]
```diff
@@ -897,9 +897,6 @@ bool AsyncRPCOperation_sendmany::find_utxos(bool fAcceptCoinbase=false) {
         destinations.insert(fromtaddr_);
     }
     vector<COutput> vecOutputs;
-
-    LOCK2(cs_main, pwalletMain->cs_wallet);
-
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/asyncrpcoperation_sendmany.cpp:902]: yeah, i didn't change it on purpose on this PR. Mainly to scope this first cleanup round over the transparent inputs flow only, then move to the unspent notes flow on a second round. But can add it here as well, it's a small thing.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 1 above)
Comment (furszy) [reply src/wallet/asyncrpcoperation_sendmany.cpp:863]: you mean to set `selectedUTXOCoinbase` to false at the beginning of the method?.

The `TxValues` object is created right before calling `find_utxos` (which is called only once per operation process), initializing `selectedUTXOCoinbase` to false by default (check the struct), then the flag is only changed here if a coinbase utxo gets selected.
Plus, `txValues` is not cached nor reused for any other operation process, dies here.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (str4d) [reply src/wallet/asyncrpcoperation_sendmany.cpp:902]: I'm fine with this being done in a subsequent PR.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -3209,7 +3209,7 @@ CAmount CWallet::GetImmatureWatchOnlyBalance() const
     return nTotal;
 }
 
-void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const CCoinControl *coinControl, bool fIncludeZeroValue, bool fIncludeCoinBase, bool fOnlySpendable) const
+void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const CCoinControl *coinControl, bool fIncludeZeroValue, bool fIncludeCoinBase, bool fOnlySpendable, int nMinDepth) const
 {
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/wallet.cpp:3220]: k updating.

**Comment by furszy** [comment]
nits applied per feedback plus rebased on master.
Changes:
1) in `AvailableCoins`: `assert(nMinDepth >= 0);` at the beginning.
2) in `AvailableCoins`: now unneeded due (1) `if (nDepth < 0)` removed.
3) unneeded recursive lock removed from `find_unspent_notes`.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/asyncrpcoperation_sendmany.cpp:902]: upps, saw this comment late. i added a final commit for it already. Not a big deal to remove it from here and add it to next round anyway.

[the remark this answers is not in the document]
Comment (str4d) [reply src/wallet/asyncrpcoperation_sendmany.cpp:902]: FIne also to leave it in :)

**Review by str4d — APPROVED** [review]
utACK. This is an excellent and easily-reviewable patchset!

**Review by daira — APPROVED** [review]
utACK changes

**Comment by daira** [comment]
@zkbot r+

**Pull request merged**